src/NATS.Client/JetStream/JetStreamMsg.cs

```diff
@@ -0,0 +1,242 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Text;
+
+namespace NATS.Client.JetStream
+{
+    public sealed class JetStreamMsg : Msg
+    {
+        static readonly private byte[] ackAck = Encoding.ASCII.GetBytes("+ACK");
+	    static readonly private byte[] ackNak = Encoding.ASCII.GetBytes("-NAK");
+	    static readonly private byte[] ackProgress = Encoding.ASCII.GetBytes("+WPI");
+	    static readonly private byte[] ackTerm = Encoding.ASCII.GetBytes("+TERM");
+
+        /// <summary>
+        /// Gets the metadata associated with a JetStream message.
+        /// </summary>
+        public override MetaData Metadata { get; }
+
+
+        private IConnection Connection { get; }
+
+        // fastpath check to see if this is a valid JS message.
+        // Speed is of the essence here.
+        static internal bool IsJS(string replySubj)
+        {
+            if (replySubj == null)
+                return false;
+
+            int len = replySubj.Length;
+            if (len < 9)
+            {
+                // quick and dirty check
+                return false;
+            }
+
+            // TODO - is this valid?
+            if (replySubj[0] != '$' ||
+                replySubj[1] != 'J' ||
+                replySubj[2] != 'S' ||
+                replySubj[3] != '.' ||
+                replySubj[4] != 'A' ||
+                replySubj[5] != 'C' ||
+                replySubj[6] != 'K' ||
+                replySubj[7] != '.')
+            {
+                return false;
+            }
+
+            int i;
+            for (i = 8; i < len; i++)
+            {
+                if (replySubj[i] == '.')
+                {
+                    i++;
+                }
+            }
+            return i == 7;
+        }
+
+        // Take the reply and parse it into the metadata.
+        internal JetStreamMsg(IConnection conn, string subject, string reply, byte[]data) : base(subject, reply, data)
+        {
+            Connection = conn;
+            Metadata = new MetaData(reply);
+        }
+
+        internal JetStream CheckReply(out bool isPullMode)
+        {
+            // TODO: Implement;
+            if ((sub is JetStreamSubscription) == false)
+            {
+                // Not a Jetstream enabled connection.
+                isPullMode = false;
+                return null;
+            }
+
+            isPullMode = (sub is JetStreamPullSubscription);
+            return ((JetStreamSubscription)sub).JS;
```

Comment:
Could be simplified/.NETified some:
```suggestion
            if (sub is JetStreamSubscription jsSub)
            {
                isPullMode = jsSub is JetStreamPullSubscription;
                return jsSub.JS;
            }

            // Not a Jetstream enabled connection
            isPullMode = false;
            return null;
```

Reply:
+1